### ISSUE TITLE
ContainerOp.set_gpu_limit: "Unschedulable: 0/4 nodes are available"

### ISSUE THREAD

**Issue by JakeTheWise** [kubeflow/pipelines#2561]
**What happened:**
I'm running a pipeline with one component specified by `kfp.dsl.ContainerOp(...).set_gpu_limit(1)`. When I run this pipeline, the KFP UI displays a yellow warning banner for the component that says: `This step is in Pending state with this message: Unschedulable: 0/4 nodes are available: 4 Insufficient nvidia.com/gpu.` The component has hung like this for over 10 minutes now. There are no container logs.

This is with kubeflow v0.7, using the default gcp_iap.yaml that contains a GPU node pool, so I'm not sure why this is happening.

The denominator in `...0/4...` of the warning seems to increment each time I try running the pipeline.

EDIT: After installing the nvidia driver on my cluster, the container now successfully completes but does not use the available GPU.

EDIT 2: Running the pipeline once more yields the original error for the component: This step is in Pending state with this message: Unschedulable: 0/3 nodes are available: 3 Insufficient nvidia.com/gpu. It also seems to have spun up another gpu node pool instead of using the existing one that was created during the kubeflow deployment.

**Comment by Ark-kun** [comment]
Do you have any GPU nodes in your cluster?

@gaoning777 Can you please take a look?

**Comment by derekhh** [comment]
@JakeTheWise - I've seen a similar issue on a self-deployed Kubeflow Pipelines a few days ago.

Can you run `kubectl get daemonset --namespace=kube-system` and see if you have the `nvidia-driver-installer` daemonset on your k8s cluster?

If not I think you need to install it via: https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#installing_drivers

**Comment by JakeTheWise** [comment]
I did not have that driver installed! So I installed it; I also had to manually scale up the gpu pool from 0 to 1 in order for the install to take effect. I re-ran the pipeline and it successfully finished. However, no GPU was used in the container run. I confirmed this by listing the available GPUs in my container using `tf.python.clientdevice_lib.list_local_devices()` which returned empty; and inspecting the device placement in tensorboard, which listed all ops on CPU. I confirmed on my GKE node dashboard that one GPU is available.

**Comment by JakeTheWise** [comment]
@derekhh @Ark-kun @gaoning777 any update on this?

**Comment by Ark-kun** [comment]
Hi, Jake.
Sorry for the wait. I've forwarded this issue to the engineer who is familiar with this area. They'll try to look at it soon.

Thanks.

**Comment by IronPan** [comment]
cc @rmgogogo

**Comment by JakeTheWise** [comment]
A similar issue occurs when trying to request TPU resources via `container_op.apply(kfp.gcp.use_tpu(tpu_cores=8, tpu_resource='v2', tf_version='2.0.0'))`.

**Comment by sid027** [comment]
> I did not have that driver installed! So I installed it; I also had to manually scale up the gpu pool from 0 to 1 in order for the install to take effect. I re-ran the pipeline and it successfully finished. However, no GPU was used in the container run. I confirmed this by listing the available GPUs in my container using `tf.python.clientdevice_lib.list_local_devices()` which returned empty; and inspecting the device placement in tensorboard, which listed all ops on CPU. I confirmed on my GKE node dashboard that one GPU is available.
> 
> EDIT: Running the pipeline once more yields the original error for the component: `This step is in Pending state with this message: Unschedulable: 0/3 nodes are available: 3 Insufficient nvidia.com/gpu.` It also seems to have spun up another gpu node pool instead of using the existing one that was created during the kubeflow deployment.

I also notice it spins up a new node pool instead of the existing one. And the new node pool does not have the correct permissions and I am not able to change it.

**Comment by sid027** [comment]
Workaround I use:
- stop autoscaling on GPU node pool
- start `n` number of GPUs on the default GPU node pool
- deploy my kubeflow pipeline.

The pipeline uses the default GPU node pool now. But this is a bad way to do things. Wonder why this is the case.

**Comment by gaoning777** [comment]
@rmgogogo Would you mind taking a look
/unassign @gaoning777

**Comment by JakeTheWise** [comment]
Any update on this?

**Comment by JakeTheWise** [comment]
I tried @sid027's workaround but it gives a `ImagePullBackOff` error on the GPU step. I know kubeflow can pull and run the image because the upstream step uses the same exact image (but runs a different function) and completes successfully.

**Comment by rmgogogo** [comment]
HI Jake, would you help detail more on the reproduce-steps? 

Following steps?
1. create a cluster (any special settings? e.x. enable GPU? which zone?)
2. top autoscaling on GPU node pool
3. start n number of GPUs on the default GPU node pool
4. deploy my kubeflow pipeline. (how?)

**Comment by sid027** [comment]
@JakeTheWise Once I started a GPU manually that my kubeflow pipeline could discover and use, I switched on the autoscaling back for the pool. Now everything works as expected. It is weird!

**Comment by JakeTheWise** [comment]
@rmgogogo Like I said, I followed @sid027 's steps.

**Comment by JakeTheWise** [comment]
The hack worked on my second attempt. I think there was an error spinning up the GPU node the first time.

Is there a timeframe for resolving this GPU pool startup issue?

**Comment by JakeTheWise** [comment]
@sid027 do your GPU-enabled pipeline steps still run on a cluster when you re-enable your GPU node pool autoscaling from `0` to `n` nodes? Or do you find that you must keep at least `1` GPU node running?

**Comment by sid027** [comment]
@JakeTheWise once it works with the hack, do this:
- enable autoscaling
- set `min nodes` to `0'.
- run the pipeline

This time, a new GPU node poll will `not` spin up but rather the default GPU node pool will be used to automatically spin up an instance and train your model. It will autoscale back to `0` around 10 mins. after the training finishes.

**Comment by JakeTheWise** [comment]
@sid027 @rmgogogo Update: even though the first time worked with the hack, after re-initalizing autoscaling and scaling the default GPU node pool back to 0, upon the next pipeline execution a new GPU node pool did spin up, and the ImagePullBackOff problem happened again.

**Comment by sid027** [comment]
@JakeTheWise is your `Node auto-provisioning` set to `Disabled`?
If you enable this, then new node pools are allowed. You can check this on the main cluster on the UI.

**Comment by JakeTheWise** [comment]
@sid027 @rmgogogo Once I disabled node auto-provisioning, it seemed to work. However, when I switched to a standalone deployment of Kubeflow Pipelines v0.2, @sid027's trick no longer works.

In addition, on both kubeflow 0.7 and kubeflow pipelines 0.2, if I turn autoscaling on for my GPU pool, it does not autoscale when I run 10 gpu-requesting pipelines in parallel.

**Comment by JakeTheWise** [comment]
@sid027 @rmgogogo I'd really like to solve this issue in my KFP standalone (now v2.1) deployment. The same exact code and pipeline works in a different cluster with kubeflow 0.7. But even in my standalone KFP cluster, the `set_gpu_limit` does not work. The cluster has a gpu node pool with nodes=1 and autoscaling disabled, but the GPU step will not run.

**Comment by sid027** [comment]
@JakeTheWise I am not sure if one of the PRs solves this issue. If not, maybe you should ask for a branch?

**Comment by JakeTheWise** [comment]
I don't see any open PRs that mention this... I'd like for others to try to recreate this issue before I open a branch

**Comment by sid027** [comment]
@JakeTheWise There is another issue because of which the scaling doesn't work sometimes - `ZONE_RESOURCE_POOL_EXHAUSTED_WITH_DETAILS`. This just means the GCP doesn't have enough hardware in that zone. I am observing this happening a lot since January 21-22 onward.

**Comment by JakeTheWise** [comment]
@sid027 Fair point, and I've hit that issue when manually scaling my GPU pools before, especially with K80 GPUs (GCP default GPU, makes sense). So as a separate issue, does kubeflow alert when this is the case?

Have you been able to recreate the autoscaling issue?

**Comment by sid027** [comment]
@JakeTheWise unfortunately not. I am able to still automatically scale if the zone has available hardware. However, I have noticed other major issues.
- I cannot run 2 pipelines at the same time. This is weird that the cluster doesn't just scale up to allow resources. As of now, the second pipeline waits until the first one stops then starts. This could also be because the second pipeline waits for the PVCs to be freed since both pipelines are expected to share underlying hard disk - need to test though!

**Comment by rmgogogo** [comment]
https://github.com/kubeflow/pipelines/pull/3484
Please help check whether the samples in this PR helps. I tested it and didn't find issue.

One possible reason I guess is that please check whether your cluster's region has the supported GPU type.

Also please check your GPU/CPU quota to see if autoscale is stopped by quota.

The title "unschedulable: 0/4..." warning msg is just warning to let you wait for more time. In my side, I normally wait for <5m to get provisioned and <10m to get released. I hit one time that when creating the P100 node-pool, I waited for around 20m.

**Comment by shotarok** [comment]
I’m wondering if the taints GKE assigns to GPU nodes prevents pods from scheduling. Please take a look at https://github.com/kubeflow/pipelines/pull/3671.

**Comment by rmgogogo** [comment]
a sample here:
https://github.com/kubeflow/pipelines/blob/master/samples/tutorials/gpu/gpu.ipynb
